File: src/war3map/ExtensionFunctions.lua
local ____exports = {}
function ____exports.IsWalkable(self, x, y)
    return not IsTerrainPathable(x, y, PATHING_TYPE_WALKABILITY)
end
function ____exports.GetRandomLocInRectUnitSafe(self, endRect)
    return Location(GetRandomReal(GetRectMinX(endRect) + 32, GetRectMaxX(endRect) - 32), GetRandomReal(GetRectMinY(endRect) + 32, GetRectMaxY(endRect) - 32))
end
function ____exports.GetTerrainHeight(self, x, y)
    local loc = Location(x, y)
    local retVar = GetLocationZ(loc)
    RemoveLocation(loc)
    return retVar
end
function ____exports.IsPointWalkable(self, p)
    return ____exports.IsWalkable(nil, p.x, p.y)
end
function ____exports.GetNumOfPlayerUnits(self, target)
    local units = CreateGroup()
    GroupEnumUnitsOfPlayer(units, target, nil)
    local foundUnit = FirstOfGroup(units)
    local count = 0
    while foundUnit ~= nil do
        foundUnit = FirstOfGroup(units)
        GroupRemoveUnit(units, foundUnit)
        count = count + 1
    end
    DestroyGroup(units)
    return count
end
return ____exports

File: war3map/ExtensionFunctions.ts
import {Point} from "./Generic/Point";

export function GetRandomLocInRectUnitSafe(endRect: rect) {
    return Location(
        GetRandomReal(GetRectMinX(endRect) + 32, GetRectMaxX(endRect) - 32),
        GetRandomReal(GetRectMinY(endRect) + 32, GetRectMaxY(endRect) - 32),
    );
}

export function GetTerrainHeight(x: number, y: number) {
    let loc = Location(x, y);
    let retVar = GetLocationZ(loc);
    RemoveLocation(loc);
    return retVar;
}

export function IsPointWalkable(p: Point) {
    return IsWalkable(p.x, p.y);
}
export function IsWalkable(x: number, y: number) {
    return !(IsTerrainPathable(x, y, PATHING_TYPE_WALKABILITY))
}

export function GetNumOfPlayerUnits(target: player): number {
    const units = CreateGroup();
    GroupEnumUnitsOfPlayer(units, target, null);
    let foundUnit = FirstOfGroup(units);
    let count = 0;

    while (foundUnit != null) {
        foundUnit = FirstOfGroup(units);
        GroupRemoveUnit(units, foundUnit);
        count += 1;
    }
    DestroyGroup(units);
    return count;
}